### PR TITLE
near pointer interactions using collision

### PR DESCRIPTION
**Pull request by StephenHodgson**
# XRTK - Mixed Reality Toolkit Change Request

## Overview

Adds near interaction pointer events using a collider referenced by the pointer.

## Target of the change:

Is this enhancement for:

- Core (core framework, interfaces and definitions)

## Breaking Changes:

- `MixedRealityInputSystemProfile.PointingRaycastLayerMasks` -> `MixedRealityInputSystemProfile.PointerRaycastLayerMasks`
- `IMixedRealityFocusProvider.FocusLayerMasks` -> `IMixedRealityFocusProvider.GlobalPointerRaycastLayerMasks`
- `IMixedRealityPointer.PrioritizedLayerMasksOverride` -> `IMixedRealityPointer.PointerRaycastLayerMasksOverride`
- added `IMixedRealityPointer.NearInteractionCollider`
- added `CollisionExtensions.IsValidCollision(this Collision collision, LayerMask[] prioritizedLayerMasks)`

## Submodule Changes

- https://github.com/XRTK/SDK/pull/168

### CODE REVIEW

File: XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs [hunk 1]
```diff
@@ -52,12 +53,17 @@ public interface IMixedRealityPointer : IEqualityComparer
...
+        Collider NearInteractionCollider { get; }
+
+        /// <summary>
+        /// Has the conditions been satisfied to enable the interaction of this pointer?
         /// </summary>
         bool IsInteractionEnabled { get; }
 
         /// <summary>
-        /// The pointer's extent when raycasting.
+        /// The far interaction pointer's extent when raycasting.
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs:66]:
```suggestion
        /// The pointer's extent when raycasting.
```

File: XRTK-Core/Packages/com.xrtk.core/Runtime/Definitions/InputSystem/MixedRealityInputSystemProfile.cs [hunk 2]
```diff
@@ -68,13 +69,15 @@ public SystemType GazeProviderType
         public float PointingExtent => pointingExtent;
 
         [SerializeField]
-        [Tooltip("The LayerMasks, in prioritized order, that are used to determine the GazeTarget when raycasting.")]
-        private LayerMask[] pointingRaycastLayerMasks = { UnityEngine.Physics.DefaultRaycastLayers };
+        [Tooltip("The Physics Layers, in prioritized order, that are used to determine the pointers target when raycasting.")]
+        [FormerlySerializedAs("pointingRaycastLayerMasks")]
+        [FormerlySerializedAs("farInteractionRaycastLayerMasks")]
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Runtime/Definitions/InputSystem/MixedRealityInputSystemProfile.cs:74]:
```suggestion
```

**Comment by StephenHodgson** [comment]
@FejZa what are we missing for this again?

**Comment by FejZa** [comment]
Well depends on what your intention with this branch is overall. But from my POV this branch is complete. The collision detection works and the events are firing. 

Everything else I do is hands related I guess.

**Comment by StephenHodgson** [comment]
Agreed, if we need anything else we'll just open another PR or put it in the other one you're working on

**Pull request merged**